mp4parse_capi/src/lib.rs

```diff
@@ -385,61 +391,99 @@ impl Read for Mp4parseIo {
-        return std::ptr::null_mut();
-    }
-    let parser = Box::new(Mp4parseParser {
-        context: MediaContext::new(),
-        io: (*io).clone(),
-        poisoned: false,
-        opus_header: HashMap::new(),
-        pssh_data: Vec::new(),
-        sample_table: HashMap::new(),
-        audio_track_sample_descriptions: HashMap::new(),
-        video_track_sample_descriptions: HashMap::new(),
-    });
-
-    Box::into_raw(parser)
+pub unsafe extern fn mp4parse_new(io: *const Mp4parseIo, status_out: *mut Mp4parseStatus) -> *mut Mp4parseParser {
+    mp4parse_new_common(io, status_out)
 }
 
 /// Allocate an `Mp4parseAvifParser*` to read from the supplied `Mp4parseIo`.
 ///
+/// See mp4parse_new; this function is identical except that it returns a
+/// pointer to an `Mp4parseAvifParser`, which (when non-null) must be paired
+/// with a call to mp4parse_avif_free.
+///
 /// # Safety
 ///
-/// This function is unsafe because it dereferences the io pointer given to it.
-/// The caller should ensure that the `Mp4ParseIo` struct passed in is a valid
-/// pointer. The caller should also ensure the members of io are valid: the
-/// `read` function should be sanely implemented, and the `userdata` pointer
-/// should be valid.
+/// Same as mp4parse_new.
 #[no_mangle]
-pub unsafe extern fn mp4parse_avif_new(io: *const Mp4parseIo) -> *mut Mp4parseAvifParser {
-    if io.is_null() || (*io).userdata.is_null() {
-        return std::ptr::null_mut();
+pub unsafe extern fn mp4parse_avif_new(io: *const Mp4parseIo, status_out: *mut Mp4parseStatus) -> *mut Mp4parseAvifParser {
+    mp4parse_new_common(io, status_out)
+}
+
+unsafe fn mp4parse_new_common<P: ContextParser>(io: *const Mp4parseIo, status_out: *mut Mp4parseStatus) -> *mut P {
+    // Validate arguments from C.
+    if io.is_null() || (*io).userdata.is_null() || (*io).read.is_none() || status_out.is_null() {
+        set_if_non_null(status_out, Mp4parseStatus::BadArg);
+        std::ptr::null_mut()
+    } else {
+        match mp4parse_new_common_safe(&mut (*io).clone()) {
+            Ok(parser) => {
+                set_if_non_null(status_out, Mp4parseStatus::Ok);
+                parser
+            }
+            Err(status) => {
+                set_if_non_null(status_out, status);
+                std::ptr::null_mut()
+            }
+        }
     }
-    if (*io).read.is_none() {
-        return std::ptr::null_mut();
+}
+
+unsafe fn set_if_non_null<T>(ptr: *mut T, val: T) {
+    if !ptr.is_null() {
+        *ptr = val;
     }
-    let parser = Box::new(Mp4parseAvifParser {
-        context: AvifContext::new(),
-        io: (*io).clone(),
-        poisoned: false,
-    });
+}
+
+fn mp4parse_new_common_safe<T: Read, P: ContextParser>(io: &mut T) -> Result<*mut P, Mp4parseStatus> {
+    let mut context = P::Context::default();
+
+    P::read(io, &mut context)
+        .map(|_| P::with_context(context))
+        .map(Box::new)
+        .map(Box::into_raw)
+        .map_err(Mp4parseStatus::from)
```

Comment:
I realize this functional style perhaps clashes with the approach in the rest of this library. I thought it was cleaner than the alternative, but I'm happy to change if y'all prefer the equivalent:
```rust
match P::read(io, &mut context) {
    Ok(_) => {
        let parser = Box::new(P::with_context(context));
        Ok(Box::into_raw(parser))
    }
    Err(e) => Err(e.into()),
}
```